plugins/android_boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,441 @@
+/*
+ * Copyright (C) 2022 Dylan Van Assche <me@dylanvanassche.be>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <string.h>
+
+#include "fu-android-boot-device.h"
+
+#define PROC_CMDLINE	 "/proc/cmdline"
+#define MAX_CMDLINE_SIZE 4096
+#define BOOT_SLOT_ARG	 "androidboot.slot_suffix"
+#define ABL_VERSION_ARG	 "androidboot.abl.revision"
+#define SERIAL_ARG	 "androidboot.serialno"
+#define UNKNOWN_VERSION	 "0.0.UNKNOWN"
+#define SECTOR_SIZE	 512
+
+struct _FuAndroidBootDevice {
+	FuUdevDevice parent_instance;
+	const char *label;
+	const char *boot_slot;
+	const char *uuid;
+	const char *version;
+	guint64 max_size;
+};
+
+G_DEFINE_TYPE(FuAndroidBootDevice, fu_android_boot_device, FU_TYPE_UDEV_DEVICE)
+
+static char *
+fu_android_boot_extract_cmdline_arg_value(const char *key)
+{
+	gint rc;
+	gint fd;
+	char pcmd[MAX_CMDLINE_SIZE];
+	g_auto(GStrv) args = NULL;
+	g_auto(GStrv) key_value = NULL;
+
+	fd = open(PROC_CMDLINE, O_RDONLY);
+	rc = read(fd, pcmd, MAX_CMDLINE_SIZE);
+	close(fd);
+
+	if (rc > 0) {
+		args = g_strsplit(pcmd, " ", -1);
+		for (guint i = 0; args[i] != NULL; i++) {
+			key_value = g_strsplit(args[i], "=", -1);
+
+			if (!(key_value[0] && key_value[1]))
+				continue;
+
+			if (g_strcmp0(key_value[0], key) == 0)
+				return g_strdup(key_value[1]);
+		}
+	} else
+		g_warning("Failed to read %s (%d)", PROC_CMDLINE, rc);
+
+	return NULL;
+}
+
+static void
+fu_android_boot_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+
+	if (self->boot_slot != NULL)
+		fu_string_append(str, idt, "BootSlot", self->boot_slot);
+
+	if (self->label != NULL)
+		fu_string_append(str, idt, "Label", self->label);
+
+	if (self->uuid != NULL)
+		fu_string_append(str, idt, "UUID", self->uuid);
+
+	fu_string_append_kx(str, idt, "MaxSize", self->max_size);
+}
+
+static gboolean
+fu_android_boot_device_probe(FuDevice *device, GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	GUdevDevice *udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	gboolean matches_boot_slot = FALSE;
+	g_autofree char *serial = NULL;
+	guint64 sectors = 0;
+	guint64 size = 0;
+
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_android_boot_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* set the physical ID */
+	if (!fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "block", error))
+		return FALSE;
+
+	/* extract label and check if it matches boot slot*/
+	if (g_udev_device_has_property(udev_device, "ID_PART_ENTRY_NAME")) {
+		self->label = g_udev_device_get_property(udev_device, "ID_PART_ENTRY_NAME");
+		g_debug("Partition label: %s", self->label);
```

Comment:
you don't need this, it's in to_string()

Reply:
I was wondering how I can easily read this to_string() information, is there a fwupdtool/fwupdmgr command for that? Maybe I just missed it